ETH-Merge-Bots/Milestones-in-TTD/src/agent.spec.ts

```diff
@@ -0,0 +1,257 @@
+import { BlockEvent, HandleBlock } from "forta-agent";
+import { TestBlockEvent } from "forta-agent-tools/lib/test";
+import { provideHandleBlock } from "./agent";
+import BigNumber from "bignumber.js";
+import { createFinding, createFinalFinding } from "./findings";
+import { getNumberOfDays, getAverageBlockTime, getEstimatedNumberOfBlocksUntilMerge } from "./utils";
+
+const TERMINAL_TOTAL_DIFFICULTY = new BigNumber("10000000");
+const blockCounter = 0;
+
+export const ETH_BLOCK_DATA = {
+  avgBlockTime: 10,
+  blockNumberAWeek: 10000,
+};
+
+const createMergeInfo = (
+  estimatedNumberOfDaysUntilMerge: number,
+  estimatedMergeDate: Date,
+  latestTotalDifficulty: string
+) => {
+  return {
+    estimatedNumberOfDaysUntilMerge,
+    estimatedMergeDate,
+    latestTotalDifficulty,
+    remainingDifficulty: TERMINAL_TOTAL_DIFFICULTY.minus(latestTotalDifficulty).toString(10),
+  };
+};
+
+describe("Milestones in Terminal Total Difficulty Bot", () => {
+  let handleBlock: HandleBlock;
+  let blockEvent: BlockEvent;
+  let avgBlockTime: number;
+
+  beforeEach(async () => {
+    handleBlock = provideHandleBlock(TERMINAL_TOTAL_DIFFICULTY, ETH_BLOCK_DATA, blockCounter);
+    blockEvent = new TestBlockEvent().setTimestamp(0);
+    blockEvent.block.difficulty = new BigNumber("10").toString();
+    avgBlockTime = getAverageBlockTime(ETH_BLOCK_DATA, blockEvent.block.timestamp, 0, 1);
+  });
+
+  it("it should not emit findings when there are more than 20 days until merge", async () => {
+    blockEvent.block.totalDifficulty = new BigNumber("8200000").toString();
+
+    const findings = await handleBlock(blockEvent);
+
+    expect(findings).toStrictEqual([]);
+  });
+
+  it("should emit 1 finding when there are less than 20 days and more than 15 days until merge", async () => {
+    blockEvent.block.totalDifficulty = new BigNumber("8300000").toString();
+
+    let findings = await handleBlock(blockEvent);
+
+    const estimatedNumberOfBlocksUntilMerge = getEstimatedNumberOfBlocksUntilMerge(
+      TERMINAL_TOTAL_DIFFICULTY,
+      new BigNumber(blockEvent.block.totalDifficulty),
+      new BigNumber(blockEvent.block.difficulty)
+    );
+
+    const { diffInDays: estimatedNumberOfDaysUntilMerge, estimatedMergeDate } = getNumberOfDays(
+      avgBlockTime,
+      estimatedNumberOfBlocksUntilMerge
+    );
+
+    const mergeInfo = createMergeInfo(
+      estimatedNumberOfDaysUntilMerge,
+      estimatedMergeDate,
+      blockEvent.block.totalDifficulty
+    );
+
+    expect(estimatedNumberOfDaysUntilMerge).toStrictEqual(20);
+    expect(findings).toStrictEqual([createFinding(mergeInfo)]);
+
+    findings = await handleBlock(blockEvent);
+    expect(findings).toStrictEqual([]);

```

Comment:
Passing block event again was to confirm it didn't create another alert for the same range of days, correct?

Reply:
that's right 👍